<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="quicksilver">
	<xacro:include filename="$(find quicksilver_description)/urdf/rplidar.urdf.xacro"/>
	<xacro:include filename="$(find quicksilver_description)/urdf/rim.urdf.xacro"/>
	<xacro:arg name="imu_visual"   default="false"/>
	<link name="origin_link"/>
  <material name="white">
    <color rgba="0 0 0 1"/>
  </material>

  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.75 0.75 0.03"/>
    </geometry>
    <origin rpy = "0 0 0" xyz="0 0 0.04"/>
      <material name="white"/>
    </visual>
    <inertial>
	  <origin xyz="0.0 0.0 0.04" rpy="0.0 0.0 0.0"/>
      <mass value="1"/>
      <inertia
        ixx="0.04695"
        ixy="0"
        ixz="0"
        iyy="0.04695"
        iyz="0"
        izz="0.09385"/>
      </inertial>
	  <collision>
		<geometry>
			<box size="0.75 0.75 0.03"/>
		</geometry>
	  </collision>
  </link>
  <joint name="base_joint" type="fixed">
		<parent link="origin_link" />
		<child link="base_link" />
		<origin xyz="0 0 0.01905"/>
	</joint>
	<joint name="imu_joint" type="fixed">
		<parent link="base_link"/>
		<child link="imu_link"/>
		<origin xyz="0 0.30 0.09" rpy="0 0 0"/>
	  </joint>
	
	  <link name="imu_link"/>
	
	
<xacro:rplidar
		name="rplidar"
		xyz="0 0.30 0.09"
		rpy="0.0 0.0 -2.28"
		color="1 0.84 0 1" />
  <xacro:rim_macro
		name="wheel4"
    	color="0.6 0 0"
		xyz="-0.393 0 0.016"
		rpy="1.570796326794896 0 1.5708"/>
	<xacro:rim_macro
		name="wheel3"
		color="0 0.6 0"
		xyz="0 -0.393 0.016"
		rpy="1.570796326794896 0 3.141592653589793"/>
	<xacro:rim_macro
		name="wheel2"
		color="0.6 0 0"
		xyz="0.393 0 0.016"
		rpy="1.570796326794896 0 -1.5708"/>
	<xacro:rim_macro
		name="wheel1"
		color="0 0.6 0"
		xyz="0 0.393 0.016"
		rpy="1.570796326794896 0 0"/>


	<gazebo reference="base_link">
    		<material>Gazebo/Black</material>
			<mu1>0.1</mu1>
			<mu2>0.1</mu2>
			<kp>500000.0</kp>
			<kd>10.0</kd>
			<minDepth>0.001</minDepth>
			<maxVel>0.1</maxVel>
			<fdir1>1 0 0</fdir1>
  	</gazebo>

	<gazebo reference="rim_wheel3_link">
    		<material>Gazebo/Green</material>
			<mu1>0.1</mu1>
			<mu2>0.1</mu2>
			<kp>500000.0</kp>
			<kd>10.0</kd>
			<minDepth>0.001</minDepth>
			<maxVel>0.1</maxVel>
			<fdir1>1 0 0</fdir1>
  	</gazebo>

	<gazebo reference="rim_wheel4_link">
    		<material>Gazebo/Red</material>
			<mu1>0.1</mu1>
			<mu2>0.1</mu2>
			<kp>500000.0</kp>
			<kd>10.0</kd>
			<minDepth>0.001</minDepth>
			<maxVel>0.1</maxVel>
			<fdir1>1 0 0</fdir1>
  	</gazebo>

	<gazebo reference="rim_wheel1_link">
    		<material>Gazebo/Green</material>
			<mu1>0.1</mu1>
			<mu2>0.1</mu2>
			<kp>500000.0</kp>
			<kd>10.0</kd>
			<minDepth>0.001</minDepth>
			<maxVel>0.1</maxVel>
			<fdir1>1 0 0</fdir1>
  	</gazebo>
  <gazebo reference="rim_wheel2_link">
    		<material>Gazebo/Red</material>
			<mu1>0.1</mu1>
			<mu2>0.1</mu2>
			<kp>500000.0</kp>
			<kd>10.0</kd>
			<minDepth>0.001</minDepth>
			<maxVel>0.1</maxVel>
			<fdir1>1 0 0</fdir1>
  </gazebo>
  <gazebo reference="imu_link">
    <sensor type="imu" name="imu">
      <always_on>true</always_on>
      <visualize>$(arg imu_visual)</visualize>
    </sensor>
    <material>Gazebo/FlatBlack</material>
  </gazebo>
	<gazebo>
		
		<plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
			<robotNamespace>/quicksilver</robotNamespace>
		</plugin>
	</gazebo>
	<gazebo>
		<plugin name="imu_plugin" filename="libgazebo_ros_imu.so">
		  <alwaysOn>true</alwaysOn>
		  <bodyName>imu_link</bodyName>
		  <frameName>imu_link</frameName>
		  <topicName>imu</topicName>
		  <serviceName>imu_service</serviceName>
		  <gaussianNoise>0.0</gaussianNoise>
		  <updateRate>100</updateRate>
		  <imu>
			<noise>
			  <type>gaussian</type>
			  <rate>
				<mean>0.0</mean>
				<stddev>2e-4</stddev>
				<bias_mean>0.0000075</bias_mean>
				<bias_stddev>0.0000008</bias_stddev>
			  </rate>
			  <accel>
				<mean>0.0</mean>
				<stddev>1.7e-2</stddev>
				<bias_mean>0.1</bias_mean>
				<bias_stddev>0.001</bias_stddev>
			  </accel>
			</noise>
		  </imu>
		</plugin>
	  </gazebo>
 </robot>
